cfgmgr/macsecmgr.cpp

```diff
@@ -0,0 +1,782 @@
+    }
+}
+
+void MACsecMgr::doTask(Consumer &consumer)
+{
+    SWSS_LOG_ENTER();
+
+    using TaskType = std::tuple<const std::string,const std::string>;
+    using TaskFunc = task_process_status (MACsecMgr::*)(const std::string &, const TaskArgs &);
+    const static std::map<TaskType, TaskFunc > TaskMap = {
+        { { CFG_MACSEC_PROFILE_TABLE_NAME, SET_COMMAND }, &MACsecMgr::loadProfile},
+        { { CFG_MACSEC_PROFILE_TABLE_NAME, DEL_COMMAND }, &MACsecMgr::removeProfile},
+        { { CFG_PORT_TABLE_NAME, SET_COMMAND }, &MACsecMgr::enableMACsec},
+        { { CFG_PORT_TABLE_NAME, DEL_COMMAND }, &MACsecMgr::disableMACsec},
+    };
+
+    const std::string & table_name = consumer.getTableName();
+    auto itr = consumer.m_toSync.begin();
+    while (itr != consumer.m_toSync.end())
+    {
+        task_process_status task_done = task_failed;
+        auto & message = itr->second;
+        const std::string & op = kfvOp(message);
+
+        auto task = TaskMap.find(std::make_tuple(table_name, op));
+        if (task != TaskMap.end())
+        {
+            task_done = (this->*task->second)(
+                kfvKey(message),
+                kfvFieldsValues(message));
+        }
+        else
+        {
+            SWSS_LOG_ERROR(
+                "Unknown task : %s - %s",
+                table_name.c_str(),
+                op.c_str());
+        }
+
+        if (task_done == task_need_retry)
+        {
+            SWSS_LOG_DEBUG(
+                "Task %s - %s need retry",
+                table_name.c_str(),
+                op.c_str());
+            ++itr;
+        }
+        else
+        {
+            if (task_done != task_success)
+            {
+                SWSS_LOG_WARN("Task %s - %s fail",
+                    table_name.c_str(),
+                    op.c_str());
+            }
+            else
+            {
+                SWSS_LOG_DEBUG(
+                    "Task %s - %s success",
+                    table_name.c_str(),
+                    op.c_str());
+            }
+
+            itr = consumer.m_toSync.erase(itr);
+        }
+    }
+}
+
+#define GetValue(args, name) (get_value(args, #name, name))
+
+bool MACsecMgr::MACsecProfile::update(const TaskArgs & ta)
+{
+    SWSS_LOG_ENTER();
+
+    // The following fields are optional
+    if (GetValue(ta, fallback_cak) && !GetValue(ta, fallback_ckn))
+    {
+        return false;
+    }
+    if (!GetValue(ta, enable_replay_protect))
+    {
+        enable_replay_protect = false;
+    }
+    if (!GetValue(ta, replay_window))
+    {
+        replay_window = 0;
+    }
+    if (!GetValue(ta, send_sci))
+    {
+        send_sci = true;
+    }
+    if (!GetValue(ta, rekey_period))
+    {
+        rekey_period = 0;
+    }
+
+    // The following fields are necessary
+    return GetValue(ta, priority)
+        && GetValue(ta, cipher_suite)
+        && GetValue(ta, primary_cak)
+        && GetValue(ta, primary_ckn)
+        && GetValue(ta, policy);
```

Comment:
Again although this is in line with the HLD, we can pick a default ("integrity_only" / "security"). Also I realize in the HLD it is mentioned "INTEGRITY_ONLY: All traffics, except EAPOL, will be converted to MACsec packets without encryption." This is not accurate. EAPoL-MKA packets are always integrity protected and have an ICV at the end. Integrity_only does not encrypt data packets but only appends the ICV. "Security" will encrypt as well.

Reply:
This is a good suggestion. But maybe we can add the default value in another PR. Because I think we should discuss which default value should be used.
@lguohan

Reply:
Hi, I fix it in this commit,https://github.com/Azure/sonic-swss/pull/1475/commits/cfdd68e1aae2897fd4a24d6994127cd9dfe1ed3e please feel free to review them.